File: 백준/Silver/15666. N과 M （12）/N과 M （12）.py
import sys
fi = sys.stdin.readline
sys.setrecursionlimit(10**8)

def solution():
    global n, m, arr, num
    n, m = map(int, fi().split())
    arr = [0 for _ in range(n)]
    num = list(map(int, fi().split()))
    num.sort()
    
    func(0, 0)

def func(k, st):
    if k == m:
        for i in range(m):
            print(arr[i], end=" ")
        print()
        return
    
    tmp = 0
    for i in range(st, n):
        if tmp != num[i]:
            arr[k] = num[i]
            tmp = arr[k]
            func(k + 1, i)
                
if __name__ == '__main__':
    solution()

File: 백준/Silver/15666. N과 M （12）/N과 M （12）.cc
#include <bits/stdc++.h>
using namespace std;
#define fastio cin.tie(0)->sync_with_stdio(0)
#define ll long long
#define X first
#define Y second

int n,m;
int arr[10];
int num[10];
// bool isused[10];

void func(int k, int st) {
  if(k==m) {
    for(int i=0;i<m;i++) {
      cout<<arr[i]<<' ';
    }
    cout<<'\n';
    return;
  }
  
  int tmp = 0;
  for(int i=st;i<n;i++) {
    if(tmp != num[i]){
      arr[k] = num[i];
      tmp = num[i];
      func(k + 1,i);
    }
  }
}

int main(void) {
  fastio;
  cin>>n>>m;
  for(int i=0;i<n;i++) cin>>num[i];
  sort(num, num+n);
  func(0,0);
  return 0;
}